### ISSUE TITLE
[no-non-null-assertion] less verbose escape hatch

### ISSUE THREAD

**Issue by mmkal** [typescript-eslint/typescript-eslint#4467]
`// eslint-disable-next-line @typescript-eslint/no-non-null-assertion` is a _really_ verbose way of saying "trust me this thing is defined". The noise-to-informativeness ratio is off.

One way that this could be improved: use `!!!`. e.g. disallow `foo!.slice()` as before, but explicitly allow `foo!!!.slice()`. It might look a bit weird at first, but it's

1. verbose enough that it stands out as something potentially dangerous.
2. terse enough that you can quickly type, and quickly mentally process it.
3. not something that anyone would type accidentally or without fairly careful consideration.

So I think on balance, could be a better option than the in-built suppression in many cases (and there are many enough of these cases - if typescript, and all the type definitions we used were perfect, it'd never be needed. But until that happens, it shouldn't be so painful to use an escape hatch).

It could be behind an option - the implementation could be pretty simple, here's a working wrapper for the rule:

```js
'bang-bang-bang': (() => {
  const parentDef = tseslint.rules['no-non-null-assertion']
  return {
    ...parentDef,
    meta: {
      ...parentDef.meta,
      messages: {
        ...parentDef.meta.messages,
        noNonNull: `Forbidden non-null assertion. Use "!!!" if you want to avoid a noisy lint suppression.`
      },
    },
    create: context => {
      const parentRule = parentDef.create(context)
      const isNNE = node => node && node.type === 'TSNonNullExpression'

      return {
        TSNonNullExpression: node => {
          if (isNNE(node.expression) && isNNE(node.expression.expression)) {
            return // !!!, permitted
          }
          if (isNNE(node.parent)) {
            return // probably the child of a !!!, rely on parent for linting
          }

          return parentRule.TSNonNullExpression(node)
        }
      }
    }
  }
})(),
```

I'd be happy to contribute a PR if it'd be accepted?

(apologies for the untemplated issue, I couldn't find a template for "feature request for existing rule")

**Comment by bradzacher** [comment]
While I understand the sentiment - this atyle would clash with other ***recommended*** rule [no-extra-non-null-assertion](https://typescript-eslint.io/rules/no-extra-non-null-assertion).

The point of a lint rule is not to be easy to bypass, it is to warn about problems. The verbosity of the bypass is a feature, not a bug.

The problem with adding a "quick and easy" bypass style is that it makes it, well quick and easy to bypass the lint rule. If it's easy then people will do it instead of addressing the lint rule properly.
This then just leads you into a state that is no than not having the rule at all.

Put it another way:

If the only way for me to bypass a lint rule is to go through the effort of a disable comment (which might also need an explanation comment for in some repos) - then I'm going to be motivated to fix the issue rather than just ignore it. I'm going to seriously consider if I really want to bypass the rule.

If I can instead bypass the lint rule by adding just two characters - well that's so simple that I'll just do that. Why bother fixing it when I can suppress it by just adding two more `!`?

-----

My experience with linting in one of the largest repos in the world is that a huge percentage of people are either lazy, or they are motivated by time constraints. The former take the easy out because it's easier than "fixing" the code. The latter take the easy out because it's faster.
Both types don't care about codebase rot - they just want to "get shit done".

So easy outs will be abused by people if you give it to them. (Which is the entire problem with the non-null assertion in the first place!!!)

-----

Sorry. All of that is a long rant to give background and explanation to my conclusion: I don't think this is a good thing to add to this project.

Thanks for the issue and discussion! Feel free to fork or create an extension rule locally for your project (or published to npm) if you want to use this style!

**Comment by mmkal** [comment]
That's all totally fair. We currently do have just such a rule in our team repo, it works nicely. Just thought I'd propose it being promoted so we don't have to maintain it.

And yeah, we had to turn off no-extra assertion.

I do agree with what you're saying, but my hope is that `!!!` will strike a balance between not too painful to read or write, but when people _do_ reach for it lazily, it still sticks out pretty clearly in code review. It's useful for the cases where the tooling is insufficient (e.g. null checking a non-readonly property, then using it in an array `.map` callback). For me, there are too many cases like that, and you end up with either tons of suppressions, which makes people start to ignore other, more important suppressions, or unnecessary null checks for the sake of the compiler, or `as` casts (or even casts to `any`). All of the above are decently risky and make the code worse.

Anyway, counter-rant over, we'll stick with our wrapper unless you change your mind!
